optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java

```diff
@@ -0,0 +1,120 @@
+        if (bestSolutionConsumer == null) {
+            throw new IllegalStateException("No best solution consumer has been defined.");
+        }
+        bestSolutionWaitingForConsumption.set(bestSolution);
+        tryConsumeWaitingIntermediateBestSolution();
+    }
+
+    // Called on the Solver thread after Solver#solve() returns.
+    void consumeFinalBestSolution(Solution_ finalBestSolution) {
+        if (finalBestSolutionConsumer == null) {
+            throw new IllegalStateException("No final best solution consumer has been defined.");
+        }
+        try {
+            // Wait for the previous consumption to complete.
+            // As the solver has already finished, holding the solver thread is not an issue.
+            activeConsumption.acquire();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException("Interrupted when waiting for the final best solution consumption.");
+        }
+        // Make sure the final best solution is consumed by the intermediate best solution consumer first.
+        // Situation:
+        // The consumer is consuming the last but one best solution. The final best solution is waiting for the consumer.
+        scheduleIntermediateBestSolutionConsumption();
+        consumerExecutor.submit(() -> {
+            try {
+                finalBestSolutionConsumer.accept(finalBestSolution);
+            } catch (Throwable throwable) {
+                exceptionHandler.accept(problemId, throwable);
+            } finally {
+                activeConsumption.release();
+            }
+        });
+    }
+
+    // Called both on the Solver thread and the Consumer thread.
+    private void tryConsumeWaitingIntermediateBestSolution() {
+        if (bestSolutionWaitingForConsumption.get() == null) {
+            return; // There is no best solution to consume.
+        }
+        if (activeConsumption.tryAcquire()) {
+            scheduleIntermediateBestSolutionConsumption().thenRunAsync(this::tryConsumeWaitingIntermediateBestSolution,
+                    consumerExecutor);
+        }
+    }
+
+    /**
+     * Called both on the Solver thread and the Consumer thread.
+     * Don't call without locking, otherwise multiple consumptions may be scheduled.
+     */
+    private CompletableFuture<Void> scheduleIntermediateBestSolutionConsumption() {
+        return CompletableFuture.runAsync(() -> {
+            Solution_ bestSolution = bestSolutionWaitingForConsumption.getAndSet(null);
+            try {
+                if (bestSolution != null) {
+                    bestSolutionConsumer.accept(bestSolution);
+                }
+            } catch (Throwable throwable) {
+                exceptionHandler.accept(problemId, throwable);
+            } finally {
+                activeConsumption.release();
+            }
+        }, consumerExecutor);
+    }
+
+    @Override
+    public void close() {
```

Comment:
I am concerned that we may have introduced a thread leak here:

It appears that this close method may not always be invoked when the solving terminates which leaves open threads in the consumerExecutor. These open threads are preventing the JVM from exiting.

It seems as if we're expecting the DefaultSolverManager, upon closing, to close the registered DefaultSolverJobs which should close the ConsumerSupport which would shut down the consumerExecutor threads.

However, when the DefaultSolverJob.call() method is finishing, the finally clause calls the DefaultSolverJob.solvingTerminated() method which _unregisters_ the solverJob from the SolverManager. Later, when the SolverManager.close() method gets called to clean up resources, since the solverJob has been unregistered, the transitive chain of close methods does not get called.

Reply:
@rsynek FYI

Reply:
@brewerc thank you for reporting this; your analysis is correct. As I think about it, there is one related issue: the `SolverManager` may run over a long period of time and the number of consumer threads will grow anyway, as nobody calls `SolverManager.close()`. Thus, the `SolverJob.close()` must be called right after the last best solution is consumed.

https://issues.redhat.com/browse/PLANNER-2708